File: android/app/src/main/java/com/musicstory/app/data/model/TrackInfo.kt
package com.musicstory.app.data.model

import com.musicstory.app.media.MediaJunkFilter

data class TrackInfo(
    val artist: String,
    val title: String,
    val album: String? = null,
    val packageName: String? = null,
    val durationMs: Long = 0L,
) {
    val displayKey: String
        get() = "${artist.lowercase()}|${title.lowercase()}"

    fun isPlaceholder(): Boolean {
        val a = artist.lowercase()
        val t = title.lowercase()
        if (a.contains("вспоминаем трек") || a.contains("remember")) return true
        if (t.contains("скоро начн") || t.contains("остановились") || t.contains("will begin")) return true
        if (t.contains("музыка скоро") || t == "paused") return true
        if (MediaJunkFilter.isJunkTrack(packageName, artist, title)) return true
        return false
    }

    fun isValid(): Boolean =
        artist.isNotBlank() &&
            title.isNotBlank() &&
            !isPlaceholder() &&
            artist.length <= MAX_FIELD_LEN &&
            title.length <= MAX_FIELD_LEN

    companion object {
        private const val MAX_FIELD_LEN = 200
    }
}

File: ios/MusicStory/Media/TrackInfo.swift
import Foundation

enum TrackSource: String, Codable, Sendable {
    case spotify
    case appleMusic
    case shazam
    case manual
}

struct TrackInfo: Equatable, Sendable {
    let artist: String
    let title: String
    let album: String?
    let source: TrackSource
    let durationMs: Int64

    init(
        artist: String,
        title: String,
        album: String? = nil,
        source: TrackSource,
        durationMs: Int64 = 0
    ) {
        self.artist = artist.trimmingCharacters(in: .whitespacesAndNewlines)
        self.title = title.trimmingCharacters(in: .whitespacesAndNewlines)
        self.album = album?.trimmingCharacters(in: .whitespacesAndNewlines)
        self.source = source
        self.durationMs = durationMs
    }

    var displayKey: String {
        "\(artist.lowercased())|\(title.lowercased())"
    }

    func isPlaceholder() -> Bool {
        let a = artist.lowercased()
        let t = title.lowercased()
        if a.contains("вспоминаем трек") || a.contains("remember") { return true }
        if t.contains("скоро начн") || t.contains("остановились") || t.contains("will begin") { return true }
        if t.contains("музыка скоро") || t == "paused" { return true }
        if MediaJunkFilter.isNonMusicPlaybackMetadata(artist: artist, title: title) { return true }
        return false
    }

    func isValid() -> Bool {
        !artist.isEmpty &&
            !title.isEmpty &&
            !isPlaceholder() &&
            artist.count <= 200 &&
            title.count <= 200
    }
}
